x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticMultiMatchQueryRewriteInterceptor.java

```diff
@@ -0,0 +1,487 @@
+        for (Map.Entry<String, Map<String, InferenceFieldMetadata>> entry : inferenceFieldsPerIndex.entrySet()) {
+            String indexName = entry.getKey();
+            Map<String, InferenceFieldMetadata> indexInferenceFields = entry.getValue();
+
+            for (String fieldName : indexInferenceFields.keySet()) {
+                SemanticQueryBuilder semanticQuery = createSemanticQuery(fieldName, queryValue, fieldsBoosts, true);
+
+                BoolQueryBuilder indexSpecificQuery = new BoolQueryBuilder();
+                indexSpecificQuery.must(semanticQuery);
+                indexSpecificQuery.filter(new TermsQueryBuilder(IndexFieldMapper.NAME, List.of(indexName)));
+
+                if (parentQuery instanceof DisMaxQueryBuilder) {
+                    ((DisMaxQueryBuilder) parentQuery).add(indexSpecificQuery);
+                } else if (parentQuery instanceof BoolQueryBuilder) {
+                    ((BoolQueryBuilder) parentQuery).should(indexSpecificQuery);
+                }
+            }
+        }
+    }
+
+    /**
+     * Adds non-inference queries for non-inference fields per index to the provided query builder.
+     */
+    private void addNonInferenceQueriesPerIndex(
+        QueryBuilder parentQuery,
+        MultiFieldInferenceInfo inferenceInfo,
+        MultiMatchQueryBuilder originalQuery,
+        Map<String, Float> fieldsBoosts
+    ) {
+        Map<String, Set<String>> nonInferenceFieldsPerIndex = inferenceInfo.nonInferenceFieldsPerIndex();
+        for (Map.Entry<String, Set<String>> entry : nonInferenceFieldsPerIndex.entrySet()) {
+            String indexName = entry.getKey();
+            Set<String> nonInferenceFields = entry.getValue();
+
+            MultiMatchQueryBuilder indexSpecificQuery = createNonInferenceQueryForIndex(originalQuery, nonInferenceFields, fieldsBoosts);
+
+            BoolQueryBuilder indexFilteredQuery = new BoolQueryBuilder();
+            indexFilteredQuery.must(indexSpecificQuery);
+            indexFilteredQuery.filter(new TermsQueryBuilder(IndexFieldMapper.NAME, List.of(indexName)));
+
+            if (parentQuery instanceof DisMaxQueryBuilder) {
+                ((DisMaxQueryBuilder) parentQuery).add(indexFilteredQuery);
+            } else if (parentQuery instanceof BoolQueryBuilder) {
+                ((BoolQueryBuilder) parentQuery).should(indexFilteredQuery);
+            }
+        }
+    }
+
+    private QueryBuilder buildBestFieldsSemanticQuery(
+        MultiMatchQueryBuilder originalQuery,
+        Map<String, Float> fieldsBoosts,
+        Set<String> inferenceFields,
+        String queryValue
+    ) {
+        DisMaxQueryBuilder disMaxQuery = QueryBuilders.disMaxQuery();
+        for (String fieldName : inferenceFields) {
+            disMaxQuery.add(createSemanticQuery(fieldName, queryValue, fieldsBoosts, false));
+        }
+        // Apply tie_breaker if specified
+        if (originalQuery.tieBreaker() != null) {
```

Comment:
We need to make sure to adhere to the defaults in `MultiMatchQueryBuilder.Type` here if they're not specified.